gen/field.go

```diff
@@ -179,9 +182,25 @@ func compileJSONTag(f *compile.FieldSpec, name string, opts ...string) *structta
 	// should be omitted from the encoding if the field has an empty value,
 	// defined as false, 0, a nil pointer, a nil interface value, and any empty
 	// array, slice, map, or string.
-	if (isReferenceType(f.Type) || isStructType(f.Type) || isPrimitiveType(f.Type)) &&
-		!f.Required && !t.HasOption("omitempty") {
-		t.Options = append(t.Options, "omitempty")
+	//
+	// If the field is marked with "!omitempty", then "omitempty" will not be added.
+	// If both "!omitempty" and "omitempty" are present, then "omitempty" is removed.
+	hasNotOmitempty := t.HasOption(notOmitempty)
+	hasOmitempty := t.HasOption(omitempty)
+	if (isReferenceType(f.Type) || isStructType(f.Type) || isPrimitiveType(f.Type)) && !f.Required {
+		if !hasNotOmitempty && !hasOmitempty {
+			t.Options = append(t.Options, omitempty)
+		} else if hasNotOmitempty && hasOmitempty {
+			newOptions := make([]string, 0)
```

Comment:
Let's add an explanation.
Also, we can pre-allocate the slice to the desired capacity.

```suggestion
		if !hasNotOmitempty && !hasOmitempty {
			// Add omitempty if it's not there and the user
			// has not opted out with !omitempty.
			t.Options = append(t.Options, omitempty)
		} else if hasNotOmitempty && hasOmitempty {
			// !omitempty takes precedence so remove omitempty
			// if present.
			newOptions := make([]string, 0, len(t.Options))
```